### ISSUE TITLE
Add type hinting to functions

### ISSUE THREAD

**Issue by f0lie** [d2l-ai/d2l-en#2571]
Sometimes it's hard to understand the intent of the code without type hints. It would help a lot in understand the the pieces of code that is added to classes.

It would likely take a while to migrate all functions to type hinting, but it's still worth discussing it in a bug.

**Comment by hiepdang-ml** [comment]
Agree! I found the book has great content but the code was poorly written by not following coding standards. No docstrings, no code comments, no type hints, poorly named variables. I hope they will rework on the code in later versions.
